File: sdk/java/src/main/java/io/forma/SidecarClient.java
package io.forma;

import java.io.*;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.UnixDomainSocketAddress;
import java.nio.channels.SocketChannel;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * HTTP client to the forma-sidecar local listener — unix domain socket
 * (default) or localhost TCP.
 *
 * Uses raw sockets (like Python's _UnixHTTPConnection) instead of
 * java.net.http.HttpClient, because HttpClient does not natively support
 * Unix domain sockets without custom SelectorProvider plumbing.
 */
public final class SidecarClient {
    private final String socketPath;
    private final String tcpHost;
    private final int tcpPort;
    private final int timeoutMillis;

    /**
     * @param endpoint "unix:///var/run/forma/sidecar.sock" or "http://localhost:PORT"
     */
    public SidecarClient(String endpoint) {
        this(endpoint, 30_000);
    }

    /**
     * @param endpoint      "unix:///var/run/forma/sidecar.sock" or "http://localhost:PORT"
     * @param timeoutMillis request timeout in milliseconds
     */
    public SidecarClient(String endpoint, int timeoutMillis) {
        this.timeoutMillis = timeoutMillis;
        if (endpoint.startsWith("unix://")) {
            this.socketPath = endpoint.substring("unix://".length());
            this.tcpHost = null;
            this.tcpPort = -1;
        } else if (endpoint.startsWith("http://")) {
            this.socketPath = null;
            var stripped = endpoint.replaceAll("^http://", "").replaceAll("/$", "");
            var colon = stripped.lastIndexOf(':');
            if (colon < 0) {
                this.tcpHost = stripped;
                this.tcpPort = 80;
            } else {
                this.tcpHost = stripped.substring(0, colon);
                this.tcpPort = Integer.parseInt(stripped.substring(colon + 1));
            }
        } else {
            throw new FormaException(
                    "sidecar endpoint " + endpoint + ": unsupported scheme (want unix:// or http://)");
        }
    }

    /**
     * POST to a sidecar endpoint.
     *
     * @param path the request path (e.g., "/ctx/db/query")
     * @param body the JSON-serializable request body
     * @return decoded JSON response as a Map
     */
    public Map<String, Object> post(String path, Map<String, Object> body) {
        var jsonBody = JsonCodec.encode(body);
        var request = buildHttpRequest("POST", path, jsonBody);

        try (var sock = createSocket()) {
            var os = new BufferedOutputStream(sock.getOutputStream());
            var is = new BufferedInputStream(sock.getInputStream());

            os.write(request.getBytes("UTF-8"));
            os.flush();

            return parseHttpResponse(is);
        } catch (FormaException e) {
            throw e;
        } catch (Exception e) {
            throw new FormaException("sidecar call " + path + ": " + e.getMessage(), e);
        }
    }

    private Socket createSocket() throws IOException {
        if (socketPath != null) {
            // Java 16+ Unix domain socket
            var addr = UnixDomainSocketAddress.of(Path.of(socketPath));
            var ch = SocketChannel.open(addr);
            var sock = ch.socket();
            sock.setSoTimeout(timeoutMillis);
            return sock;
        } else {
            var sock = new Socket();
            sock.connect(new InetSocketAddress(tcpHost, tcpPort), timeoutMillis);
            sock.setSoTimeout(timeoutMillis);
            return sock;
        }
    }

    private String buildHttpRequest(String method, String path, String body) {
        var sb = new StringBuilder();
        sb.append(method).append(" ").append(path).append(" HTTP/1.1\r\n");
        sb.append("Host: localhost\r\n");
        sb.append("Content-Type: application/json\r\n");
        sb.append("Content-Length: ").append(body.getBytes().length).append("\r\n");
        sb.append("Connection: close\r\n");
        sb.append("\r\n");
        sb.append(body);
        return sb.toString();
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> parseHttpResponse(InputStream is) throws IOException {
        // Parse status line
        var statusLine = readLine(is);
        if (statusLine == null || statusLine.isEmpty()) {
            throw new FormaException("empty response from sidecar");
        }
        var parts = statusLine.split(" ", 3);
        int statusCode = Integer.parseInt(parts[1]);

        // Parse headers (skip them)
        int contentLength = -1;
        String line;
        while ((line = readLine(is)) != null && !line.isEmpty()) {
            var lower = line.toLowerCase();
            if (lower.startsWith("content-length:")) {
                contentLength = Integer.parseInt(line.substring(line.indexOf(':') + 1).trim());
            }
        }

        // Parse body
        String body;
        if (contentLength >= 0) {
            var buf = new byte[contentLength];
            int off = 0;
            while (off < contentLength) {
                int n = is.read(buf, off, contentLength - off);
                if (n < 0) break;
                off += n;
            }
            body = new String(buf, 0, off, "UTF-8");
        } else {
            // Chunked or no length — read until connection close
            var baos = new ByteArrayOutputStream();
            byte[] buf = new byte[4096];
            int n;
            while ((n = is.read(buf)) >= 0) {
                baos.write(buf, 0, n);
            }
            body = baos.toString("UTF-8");
        }

        var decoded = JsonCodec.decode(body);
        if (statusCode != 200) {
            var errMsg = decoded != null
                    ? (String) decoded.getOrDefault("error", "HTTP " + statusCode)
                    : "HTTP " + statusCode;
            throw new FormaException("sidecar call returned " + statusCode + ": " + errMsg);
        }
        return decoded != null ? decoded : new LinkedHashMap<>();
    }

    private String readLine(InputStream is) throws IOException {
        var baos = new ByteArrayOutputStream();
        int b;
        while ((b = is.read()) >= 0) {
            if (b == '\r') {
                int next = is.read();
                if (next != '\n' && next >= 0) {
                    baos.write(next);
                }
                break;
            }
            if (b == '\n') break;
            baos.write(b);
        }
        return baos.size() > 0 ? baos.toString("UTF-8") : null;
    }
}

File: sdk/dotnet/src/FormaLib/SidecarClient.cs
using System.Net.Sockets;
using System.Text;

namespace Forma;

/// <summary>
/// HTTP client to the forma-sidecar local listener — unix domain socket
/// (default) or localhost TCP.
/// </summary>
internal sealed class SidecarClient
{
    private readonly string? _socketPath;
    private readonly string? _tcpHost;
    private readonly int _tcpPort;
    private readonly int _timeoutMillis;

    /// <param name="endpoint">"unix:///tmp/forma/sidecar.sock" or "http://localhost:PORT"</param>
    public SidecarClient(string endpoint, int timeoutMillis = 30_000)
    {
        _timeoutMillis = timeoutMillis;
        if (endpoint.StartsWith("unix://"))
        {
            _socketPath = endpoint["unix://".Length..];
        }
        else if (endpoint.StartsWith("http://"))
        {
            var stripped = endpoint["http://".Length..].TrimEnd('/');
            var colon = stripped.LastIndexOf(':');
            if (colon < 0)
            {
                _tcpHost = stripped;
                _tcpPort = 80;
            }
            else
            {
                _tcpHost = stripped[..colon];
                _tcpPort = int.Parse(stripped[(colon + 1)..]);
            }
        }
        else
        {
            throw new FormaException(
                $"sidecar endpoint {endpoint}: unsupported scheme (want unix:// or http://)");
        }
    }

    public Dictionary<string, object?> Post(string path, Dictionary<string, object?> body)
    {
        var jsonBody = JsonCodec.Encode(body);
        var request = BuildHttpRequest("POST", path, jsonBody);

        using var socket = CreateSocket();
        socket.Connect(GetEndPoint());
        socket.Send(Encoding.UTF8.GetBytes(request));
        return ParseHttpResponse(socket);
    }

    private Socket CreateSocket()
    {
        if (_socketPath is not null)
        {
            return new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
        }
        return new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp)
        {
            ReceiveTimeout = _timeoutMillis,
            SendTimeout = _timeoutMillis
        };
    }

    private System.Net.EndPoint GetEndPoint()
    {
        if (_socketPath is not null)
        {
            return new UnixDomainSocketEndPoint(_socketPath);
        }
        var addr = System.Net.Dns.GetHostAddresses(_tcpHost!)[0];
        return new System.Net.IPEndPoint(addr, _tcpPort);
    }

    private static string BuildHttpRequest(string method, string path, string body)
    {
        var sb = new StringBuilder();
        sb.Append(method).Append(' ').Append(path).Append(" HTTP/1.1\r\n");
        sb.Append("Host: localhost\r\n");
        sb.Append("Content-Type: application/json\r\n");
        sb.Append("Content-Length: ").Append(Encoding.UTF8.GetByteCount(body)).Append("\r\n");
        sb.Append("Connection: close\r\n");
        sb.Append("\r\n");
        sb.Append(body);
        return sb.ToString();
    }

    private static Dictionary<string, object?> ParseHttpResponse(Socket socket)
    {
        using var ns = new NetworkStream(socket, ownsSocket: true);
        using var reader = new StreamReader(ns, Encoding.UTF8);

        // Parse status line
        var statusLine = reader.ReadLine();
        if (string.IsNullOrEmpty(statusLine))
            throw new FormaException("empty response from sidecar");

        var parts = statusLine.Split(' ', 3);
        var statusCode = int.Parse(parts[1]);

        // Parse headers
        int contentLength = -1;
        string? line;
        while (!string.IsNullOrEmpty(line = reader.ReadLine()))
        {
            var lower = line.ToLowerInvariant();
            if (lower.StartsWith("content-length:"))
            {
                contentLength = int.Parse(line[(line.IndexOf(':') + 1)..].Trim());
            }
        }

        // Parse body
        string body;
        if (contentLength >= 0)
        {
            var buf = new char[contentLength];
            var totalRead = 0;
            while (totalRead < contentLength)
            {
                var n = reader.Read(buf, totalRead, contentLength - totalRead);
                if (n <= 0) break;
                totalRead += n;
            }
            body = new string(buf, 0, totalRead);
        }
        else
        {
            body = reader.ReadToEnd();
        }

        var decoded = JsonCodec.Decode(body);
        if (statusCode != 200)
        {
            var errMsg = decoded?.GetValueOrDefault("error") is string err
                ? err
                : $"HTTP {statusCode}";
            throw new FormaException($"sidecar call {path}: {errMsg}");
        }

        return decoded ?? new();
    }
}
